typings/connection.d.ts

```diff
@@ -493,9 +494,17 @@ export interface EventContext {
    */
   sender?: Sender;
   /**
-   * @property {Error} error An optional error object.
+   * @property {Error | ConnectionError} [error] An optional error object.
+   * - On `connection_error` event this property will be present. It will have the same information as
+   * `connection.error` but the type will be `ConnectionError`.
+   * - An error with SASL will be available through this property, but not through `connection.error`
+   * (as the amqp connection was never established).
+   * - On `disconnected` event the context will have an error property that will be of type
+   * `Error` (or some subclass) as emitted by the underlying socket.
+   * - The `session_error`, `sender_error` and `receiver_error` events will not have this (`error`)
+   * property on the EventContext.
    */
-  error?: Error;
+  error?: Error | ConnectionError;
```

Comment:
ConnectionError is a subtype of Error, my instinct would be to leave the type as is.

Reply:
I see. TypeScript supports type unions whereas other strongly typed languages like C#, Java, etc. don't. So in cases where the consumer of the API is sure that return type is `ConnectionError` then they can cast it as ConnectionError and use the object accordingly. (You would do this in C# or Java as well). The only benefit of a Union type is discoverability (which is what TS type definitions are trying to solve). Otherwise, it would be difficult for developers to find out that in certain cases this could be a ConnectionError.

Hence I added the union type for the error property on the context and in the return type over here.

Actually, just noticed something, I feel definition of get_error() actually either returns a `ConnectionError | undefined`.

```js
Connection.prototype.get_error = function () {
    if (this.transport_error) return this.transport_error; // ConnectionError
    if (this.remote.close && this.remote.close.error) { // this case returns ConnectionError
        return new errors.ConnectionError(this.remote.close.error.description, this.remote.close.error.condition, this);
    }
    return undefined; // base case
};
```

and `transport_error` is being set only at one place inside the sasl_failed method, where it is being set to an instance of ConnectionError object
```js
Connection.prototype.sasl_failed = function (text) {
    this.transport_error = new errors.ConnectionError(text, 'amqp:unauthorized-access', this);
    this._handle_error();
};
```

Reply:
If you feel the union is better, go with that. You are also right that at present get_error() will always be a ConnectionError.